File: scripts/load_test.js
#!/usr/bin/env node

const DEFAULT_TEXT =
  "Careful source evaluation matters because reliable arguments depend on evidence, context, and transparent reasoning.";

function parseArgs(argv) {
  const args = {
    url: "http://127.0.0.1:8000/predict",
    requests: 50,
    text: DEFAULT_TEXT,
  };
  for (let index = 0; index < argv.length; index += 2) {
    const key = argv[index];
    const value = argv[index + 1];
    if (key === "--url") args.url = value;
    if (key === "--requests") args.requests = Number(value);
    if (key === "--text") args.text = value;
  }
  return args;
}

function percentile(values, q) {
  const ordered = [...values].sort((left, right) => left - right);
  const index = Math.min(ordered.length - 1, Math.max(0, Math.round((ordered.length - 1) * q)));
  return ordered[index];
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  const latencies = [];
  for (let index = 0; index < args.requests; index += 1) {
    const started = performance.now();
    const response = await fetch(args.url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ text: args.text }),
    });
    if (!response.ok) {
      throw new Error(`Request failed with status ${response.status}`);
    }
    await response.json();
    latencies.push((performance.now() - started) / 1000);
  }
  const median = percentile(latencies, 0.5);
  const p95 = percentile(latencies, 0.95);
  console.log(`requests: ${latencies.length}`);
  console.log(`median_seconds: ${median.toFixed(4)}`);
  console.log(`p95_seconds: ${p95.toFixed(4)}`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: scripts/load_test.py
from __future__ import annotations

import statistics
import time

import httpx


DEFAULT_TEXT = (
    "Careful source evaluation matters because reliable arguments depend on "
    "evidence, context, and transparent reasoning."
)


def percentile(values: list[float], q: float) -> float:
    if not values:
        raise ValueError("values must not be empty")
    ordered = sorted(values)
    index = min(len(ordered) - 1, max(0, round((len(ordered) - 1) * q)))
    return ordered[index]


def run(
    url: str = "http://127.0.0.1:8000/predict",
    requests: int = 50,
    timeout: float = 30.0,
    text: str = DEFAULT_TEXT,
) -> None:
    latencies: list[float] = []
    with httpx.Client(timeout=timeout) as client:
        for _ in range(requests):
            started = time.perf_counter()
            response = client.post(url, json={"text": text})
            response.raise_for_status()
            latencies.append(time.perf_counter() - started)

    print(f"requests: {len(latencies)}")
    print(f"median_seconds: {statistics.median(latencies):.4f}")
    print(f"p95_seconds: {percentile(latencies, 0.95):.4f}")


def main() -> None:
    import fire

    fire.Fire(run)


if __name__ == "__main__":
    main()
